badge support

ShayanJavadi opened the pull request:
Added badge support. This also adds react-native-material-ui as a dependency. 
<img width="259" alt="screen shot 2018-02-04 at 2 22 45 am" src="https://user-images.githubusercontent.com/11575429/35775660-92d71ee6-0952-11e8-8341-09e87021e38d.png">

https://github.com/timomeh/react-native-material-bottom-navigation/issues/52

ShayanJavadi committed:
badge support

added badge support

timomeh commented:
Thank you for your PR. While I don't have anything against dependencies, react-native-material-ui is quite a big dependency just to use their Badge.

Implementing a Badge without react-native-material-ui wouldn't be that hard, I think.

I would accept this PR without react-native-material-ui, but at the current state I can't justify shipping this with such a big dependency for one feature.

ShayanJavadi commented:
That makes sense. I will update the pr by implementing a badge and removing the dependency then. Stay tuned.

ShayanJavadi committed:
remove react-native-material-ui

removed react-native-material-ui as dependency. implemented importing local badge component.

ShayanJavadi committed:
Added Badge.js

Added Badge component to use instead of react-native-material-ui badge

ShayanJavadi committed:
updated prop types

updated prop types

ShayanJavadi committed:
removed old prop type

removed old prop type

ShayanJavadi committed:
documented badge changes

documented badge changes

ShayanJavadi committed:
fixed isBadgeVisible documentation

fixed isBadgeVisible documentation

ShayanJavadi committed:
Final changes

final changes

ShayanJavadi committed:
final changes

final changes

ShayanJavadi commented:
@timomeh I've update the pr,  implemented the Badge component, and updated the documentation to reflect the changes. Please review my changes and let me know if you have any feedback.

Here's screenshots of the badge:

<img width="252" alt="screen shot 2018-02-07 at 1 06 42 am" src="https://user-images.githubusercontent.com/11575429/35903075-42a06b00-0ba3-11e8-80be-c77d6377b2ae.png">
<img width="252" alt="screen shot 2018-02-07 at 1 06 47 am" src="https://user-images.githubusercontent.com/11575429/35903076-42b681d8-0ba3-11e8-969a-1e5d79b014d5.png">

ShayanJavadi committed:
removed extra dependencies

removed extra dependencies

timomeh commented on lib/Badge.js:
```diff
@@ -0,0 +1,83 @@
+  render() {
+    const { children, text, isVisible } = this.props;
+    const styles = this.createStyles();
+    if (!isVisible) {
```
_For android compatibility:_ Remove the isVisible check, lines 70-72. The check is already done in Tab.js

ShayanJavadi replied:
So the check `badgeText === undefined && !isBadgeVisible` is to prevent the badge from rendering when there is no text. The check in here is for the user to be able to hide a badge even when there is text. We should keep this check

timomeh replied:
Ok, I see.

timomeh commented on lib/Badge.js:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    return (
+      <View style={{ flexDirection: "row" }}>
```
_For android compatibility:_ Remove the wrapping `<View style={{ flexDirection: "row" }}>`.

ShayanJavadi replied:
Done

timomeh commented on lib/Badge.js:
```diff
@@ -0,0 +1,83 @@
+
+    return (
+      <View style={{ flexDirection: "row" }}>
+        {children ? children : null}
```
_For android compatibility:_ children (which is the Tab Text) will be rendered in Tab.js, so remove this line

ShayanJavadi replied:
Done

timomeh commented on lib/Badge.js:
```diff
@@ -0,0 +1,83 @@
+    const container = Object.assign(
+      {},
+      {
+        position: "relative",
```
_For android compatibility:_ Change to `position: 'absolute'`

timomeh replied:
And to make the `position: absolute` work, add `position: 'relative'` to Tab.js `StyleSheets.container`. Somewhere here:

https://github.com/ShayanJavadi/react-native-material-bottom-navigation/blob/3f70221a611029b13f416f6ab51c2de569679d40/lib/Tab.js#L295-L304

ShayanJavadi replied:
done

timomeh commented on lib/Badge.js:
```diff
@@ -0,0 +1,83 @@
+        justifyContent: "center",
+        backgroundColor: "#F50057",
+        zIndex: 9999,
+        bottom: 5,
```
_For android compatibility:_ Change `bottom` and `right` to something so it's correctly positioned. It worked for me with

```
top: 3,
left: '50%',
marginLeft: 15
```

ShayanJavadi replied:
done

timomeh commented on lib/Tab.js:
```diff
@@ -121,9 +126,27 @@ export default class Tab extends Component {
+        style={badgeStyle}
+        isVisible={isBadgeVisible}
+      >
+        {active && activeIcon ? activeIcon : icon}
```
_For android compatibility:_ The icon is rendered in Tab.js, so remove this line.

ShayanJavadi replied:
done

timomeh commented on lib/Tab.js:
```diff
@@ -133,8 +156,9 @@ export default class Tab extends Component {
+
         <View ref="_bnic" collapsable={false}>
-          {active && activeIcon ? activeIcon : icon}
+          {this._renderBadge()}
```
_For android compatibility:_ Change back to `{active && activeIcon ? activeIcon : icon}`. Move `{this._renderBadge()}` to the render function, right above `{this._renderIcon()}`

ShayanJavadi replied:
done. moved it below it since moving it above was causing problems on iOS.

timomeh commented on lib/Badge.js:
```diff
@@ -0,0 +1,83 @@
+  style: {},
+}
+
+export default class RippleBackgroundTransition extends Component {
```
`class RippleBackgroundTransition` -> `class Badge` :)

ShayanJavadi replied:
done

timomeh commented on lib/Tab.js:
```diff
@@ -121,9 +126,27 @@ export default class Tab extends Component {
+    const { active, icon, activeIcon, badgeText, badgeSize, badgeStyle, isBadgeVisible } = this.props;
+
+    if (badgeText === undefined && !isBadgeVisible) {
+      return active && activeIcon ? activeIcon : icon;
```
_For android compatibility:_ Change to `return null`, so nothing gets rendered if the badge is invisible.

ShayanJavadi replied:
done

timomeh reviewed (changes requested):
Thank you for your PR, I love it! ❤️

Unfortunately it doesn't render correctly in Android, you can only see a quarter of the badge. This is because in Android everything is `overflow: hidden`, and the container of the Tab icon cuts the badge.

Fortunately, you wrote pretty good code, so it's really easy to fix it and make it work in Android. I tried it locally and – instead of let you figure it out yourself – I commented the changes you need to do in order to support Android.

ShayanJavadi commented:
@timomeh You're welcome! Thank you for making this library! I will implement/test your requested changes and will update the pr soon (thanks for figuring out the android compatibility issues!).

ShayanJavadi committed:
android compatibility 

android compatibility

ShayanJavadi committed:
android compatibility

android compatibility

ShayanJavadi commented:
@timomeh PR has been updated with the requested changes. Here's screenshots from both iOS and Android:
<img width="458" alt="screen shot 2018-02-11 at 2 30 29 am" src="https://user-images.githubusercontent.com/11575429/36071420-95d2e4be-0ed3-11e8-870e-0fa3e46b6120.png">
<img width="264" alt="screen shot 2018-02-11 at 2 30 35 am" src="https://user-images.githubusercontent.com/11575429/36071421-95e1c84e-0ed3-11e8-9a1a-c92dc4439c94.png">

timomeh commented:
Looks good to me! I'm going to merge it and make a now release.
